# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## 读取数据

In [21]:
import pandas as pd

In [22]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv") 

In [23]:
original_titles

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
1       tm84618                                Taxi Driver  MOVIE   
2      tm154986                                Deliverance  MOVIE   
3      tm127384            Monty Python and the Holy Grail  MOVIE   
4      tm120801                            The Dirty Dozen  MOVIE   
...         ...                                        ...    ...   
5845  tm1014599                                  Fine Wine  MOVIE   
5846   tm898842                                C/O Kaadhal  MOVIE   
5847  tm1059008                                    Lokillo  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   

                                            description  release_year  \
0     This collection includes 12 World War II-era p...          1945   
1     A mentally unstable Vietnam War veteran works ...          1976   
2     Intent on seeing the Cahulawassee River before...          1972   
3     King Arthur, accompanied by his squire, recrui...          1975   
4     12 American military prisoners in World War II...          1967   
...                                                 ...           ...   
5845  A beautiful love story that can happen between...          2021   
5846  A heart warming film that explores the concept...          2021   
5847  A controversial TV host and comedian who has b...          2021   
5848  Jamie Foxx, David Alan Grier and more from the...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   

     age_certification  runtime                                       genres  \
0                TV-MA       51                            ['documentation']   
1                    R      114                           ['drama', 'crime']   
2                    R      109  ['drama', 'action', 'thriller', 'european']   
3                   PG       91              ['fantasy', 'action', 'comedy']   
4                  NaN      150                            ['war', 'action']   
...                ...      ...                                          ...   
5845               NaN      100                         ['romance', 'drama']   
5846               NaN      134                                    ['drama']   
5847               NaN       90                                   ['comedy']   
5848             PG-13       37                                           []   
5849               NaN        7            ['family', 'animation', 'comedy']   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
0                  ['US']      1.0         NaN         NaN         NaN   
1                  ['US']      NaN   tt0075314         8.2    808582.0   
2                  ['US']      NaN   tt0068473         7.7    107673.0   
3                  ['GB']      NaN   tt0071853         8.2    534486.0   
4            ['GB', 'US']      NaN   tt0061578         7.7     72662.0   
...                   ...      ...         ...         ...         ...   
5845               ['NG']      NaN  tt13857480         6.8        45.0   
5846                   []      NaN  tt11803618         7.7       348.0   
5847               ['CO']      NaN  tt14585902         3.8        68.0   
5848               ['US']      NaN         NaN         NaN         NaN   
5849                   []      1.0  tt13711094         7.8        18.0   

      tmdb_popularity  tmdb_score  
0               0.600         NaN  
1              40.965       8.179  
2              10.010       7.300  
3              15.461       7.811  
4              20.398       7.600  
...               ...         ...  
5845            1.466         NaN  
5846              NaN         NaN  
5847           26.005       6.300  
5848            1.296      10.000  
5849            2.289      10.000  

[5850 rows x 15 columns]

In [24]:
original_credits

person_id         id                name                character  \
0           3748    tm84618      Robert De Niro            Travis Bickle   
1          14658    tm84618        Jodie Foster            Iris Steensma   
2           7064    tm84618       Albert Brooks                      Tom   
3           3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4          48933    tm84618     Cybill Shepherd                    Betsy   
...          ...        ...                 ...                      ...   
77796     736339  tm1059008    Adelaida Buscato                María Paz   
77797     399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798     373198  tm1059008         Inés Prieto                    Fanny   
77799     378132  tm1059008        Isabel Gaona                   Cacica   
77800    1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

## 评估和清理数据

##### 在这一部分中，我们将对在上一部分建立的original_titles及original_creditsDataFrame所包含的数据进行评估和清理。
##### 主要从两个方面进行：结构和内容，即整齐度和干净度。
##### 数据的结构性问题指不符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”这三个标准；
##### 数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

##### 为了区分开经过清理的数据和原始的数据，我们创建新的变量cleaned_titles，让它为original_titles复制出的副本，
##### 以及创建新的变量cleaned_credits，让它为original_credits复制出的变量。
##### 我们之后的清理步骤都将被运用在cleaned_titles和cleaned_credits上。

In [25]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### 数据整齐度

In [26]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

##### genres和production_countries中一个变量包含多个值，要对其进行拆分

In [27]:
cleaned_titles["genres"] = cleaned_titles["genres"].apply(lambda x :eval(x))
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].apply(lambda x :eval(x))

In [28]:
cleaned_titles.sample(10)

id                                              title   type  \
1122    ts37143                                              Alone   SHOW   
629    ts124246                                        Mighty Raju   SHOW   
2511    ts81886                          All About the Washingtons   SHOW   
3253   tm451078  Pokemon the Movie: Mewtwo Strikes Back - Evolu...  MOVIE   
498     tm99450                                      Break Ke Baad  MOVIE   
4981  tm1002815                                         Nightbooks  MOVIE   
1607   tm327551                                      Born in Syria  MOVIE   
2165   tm434917                                         I Am Jonas  MOVIE   
4451   tm849599                                            Furtive  MOVIE   
3153   tm840196                         The Trial of the Chicago 7  MOVIE   

                                            description  release_year  \
1122  Hardcore survivalists are put by themselves in...          2015   
629   A four-year-old Raju receives super human stre...          2010   
2511  The wife of a legendary rapper launches her ow...          2018   
3253  After accepting an invitation from a mysteriou...          2019   
498   Aaliya is not ready for commitment even after ...          2010   
4981  Alex, a boy obsessed with scary stories, is im...          2021   
1607  This intimate documentary follows a group of S...          2016   
2165  When Jonas was 14 he met the charismatic but m...          2018   
4451  Park Ranger Pablo Silva remakes his life in hi...          2020   
3153  What was supposed to be a peaceful protest tur...          2020   

     age_certification  runtime                                genres  \
1122             TV-MA       51              [reality, documentation]   
629                NaN       20                   [animation, action]   
2511              TV-G       25                       [comedy, music]   
3253                PG       98  [animation, action, fantasy, family]   
498                NaN      120              [comedy, drama, romance]   
4981               NaN      103             [family, fantasy, horror]   
1607               NaN       86             [documentation, european]   
2165               NaN       82            [drama, romance, european]   
4451               NaN       84                            [thriller]   
3153                 R      130            [drama, history, thriller]   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
1122                 [US]      9.0   tt4803766         8.4      5037.0   
629                  [IN]      4.0   tt6491030         4.3        61.0   
2511                 [US]      1.0   tt7661368         4.5       723.0   
3253                 [JP]      NaN   tt8856470         5.7      5475.0   
498                  [IN]      NaN   tt1578261         5.1      4275.0   
4981             [CA, US]      NaN  tt10521144         5.7      8573.0   
1607             [ES, DK]      NaN   tt5356680         7.7       280.0   
2165                 [FR]      NaN   tt8168186         7.0      5896.0   
4451                 [AR]      NaN  tt11287390         5.1       328.0   
3153                 [US]      NaN   tt1070874         7.7    173103.0   

      tmdb_popularity  tmdb_score  
1122           26.930         7.0  
629             1.225         NaN  
2511            2.020         4.6  
3253          146.188         6.7  
498             3.068         5.8  
4981           80.549         6.3  
1607            2.143         7.2  
2165            8.670         6.8  
4451            1.945         6.5  
3153           15.070         7.8

In [29]:
cleaned_titles = cleaned_titles.explode("genres")#explode() 方法默认只能处理一个列
cleaned_titles

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
1       tm84618                                Taxi Driver  MOVIE   
1       tm84618                                Taxi Driver  MOVIE   
2      tm154986                                Deliverance  MOVIE   
2      tm154986                                Deliverance  MOVIE   
...         ...                                        ...    ...   
5847  tm1059008                                    Lokillo  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   

                                            description  release_year  \
0     This collection includes 12 World War II-era p...          1945   
1     A mentally unstable Vietnam War veteran works ...          1976   
1     A mentally unstable Vietnam War veteran works ...          1976   
2     Intent on seeing the Cahulawassee River before...          1972   
2     Intent on seeing the Cahulawassee River before...          1972   
...                                                 ...           ...   
5847  A controversial TV host and comedian who has b...          2021   
5848  Jamie Foxx, David Alan Grier and more from the...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                 [US]      1.0   
1                    R      114          drama                 [US]      NaN   
1                    R      114          crime                 [US]      NaN   
2                    R      109          drama                 [US]      NaN   
2                    R      109         action                 [US]      NaN   
...                ...      ...            ...                  ...      ...   
5847               NaN       90         comedy                 [CO]      NaN   
5848             PG-13       37            NaN                 [US]      NaN   
5849               NaN        7         family                   []      1.0   
5849               NaN        7      animation                   []      1.0   
5849               NaN        7         comedy                   []      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5847  tt14585902         3.8        68.0           26.005       6.300  
5848         NaN         NaN         NaN            1.296      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[15147 rows x 15 columns]

In [30]:
cleaned_titles = cleaned_titles.explode("production_countries")#explode() 方法默认只能处理一个列
cleaned_titles

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
1       tm84618                                Taxi Driver  MOVIE   
1       tm84618                                Taxi Driver  MOVIE   
2      tm154986                                Deliverance  MOVIE   
2      tm154986                                Deliverance  MOVIE   
...         ...                                        ...    ...   
5847  tm1059008                                    Lokillo  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   

                                            description  release_year  \
0     This collection includes 12 World War II-era p...          1945   
1     A mentally unstable Vietnam War veteran works ...          1976   
1     A mentally unstable Vietnam War veteran works ...          1976   
2     Intent on seeing the Cahulawassee River before...          1972   
2     Intent on seeing the Cahulawassee River before...          1972   
...                                                 ...           ...   
5847  A controversial TV host and comedian who has b...          2021   
5848  Jamie Foxx, David Alan Grier and more from the...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
1                    R      114          drama                   US      NaN   
1                    R      114          crime                   US      NaN   
2                    R      109          drama                   US      NaN   
2                    R      109         action                   US      NaN   
...                ...      ...            ...                  ...      ...   
5847               NaN       90         comedy                   CO      NaN   
5848             PG-13       37            NaN                   US      NaN   
5849               NaN        7         family                  NaN      1.0   
5849               NaN        7      animation                  NaN      1.0   
5849               NaN        7         comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5847  tt14585902         3.8        68.0           26.005       6.300  
5848         NaN         NaN         NaN            1.296      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[17818 rows x 15 columns]

##### 拆分好了

In [31]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

##### 从头部的10行数据来看，cleaned_credits数据符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”，因此不存在结构性问题。

### 数据干净度

##### 看有没有缺失数据，缺失的数据对分析有没有影响，有没有数据类型的错误，有没有重复数据

In [32]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

##### 从输出结果来看，cleaned_titles数据共有17818条观察值，
##### title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、imdb_votes、tmdb_popularity、tmdb_score变量均存在缺失值

##### release_year表示年份，数据类型不应为数字，应为日期，所以需要进行数据格式转换，
##### seasons是季数，数据类型应是整数，
##### imdb_votes是投票数，数据类型应是整数

#### 缺失值问题

##### title：影视作品标题。description：简短描述。age_certification：适龄认证。genres：流派类型列表。
##### production_countries：出品国家列表。seasons：如果是电视剧，则是季数。imdb_id：IMDB的ID。
##### imdb_score：IMDB的评分。imdb_votes：IMDB的投票数。tmdb_popularity：TMDB的流行度。tmdb_score：TMDB的评分。
##### title,description,age_certification,production_countries,seasons,imdb_id,imdb_votes,tmdb_popularity,tmdb_score信息缺失影响不大，
##### 但imdb_score和genres，即IMDB评分和流派，和我们后续要做的分析息息相关。
##### 先提取出imdb_score缺失观察值进行查看。

In [33]:
cleaned_titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description  release_year  \
0     This collection includes 12 World War II-era p...          1945   
75    Sane Man was filmed before Bill recorded ‘Dang...          1989   
145   “Hajimete no Otsukai” (First Errand) is a Japa...          1991   
145   “Hajimete no Otsukai” (First Errand) is a Japa...          1991   
145   “Hajimete no Otsukai” (First Errand) is a Japa...          1991   
...                                                 ...           ...   
5810  Two competitive social media Influencers go he...          2021   
5833  When a family run barber shop in the heart of ...          2021   
5840  In 14th-century Mali, an ambitious young royal...          2022   
5844  The stars of "Bling Empire" discuss the show's...          2021   
5848  Jamie Foxx, David Alan Grier and more from the...          2021   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

##### 发现缺少imdb_score的同时，也缺少核心数据genres，我们将把这些观察值删除

In [41]:
cleaned_titles = cleaned_titles.dropna(subset=["imdb_score"])
cleaned_titles["imdb_score"].isnull().sum()

np.int64(0)

##### 提取出genres缺失观察值进行查看。

In [39]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description  release_year  \
1813  TV legend David Letterman teams up with fascin...          2018   
1939  MInecraft: Story Mode is an interactive, anima...          2018   
2386  In this unscripted series starring comedy lege...          2018   
2658  #ABtalks is a YouTube interview show hosted by...          2018   
4274                                                NaN          2020   
4648  'In Vitro' is an otherworldly rumination on me...          2019   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

##### 里面还存在空字符串表示的流派，并非有效数据，因此可以进行删除。删除后，查看`cleaned_titles`里是否还存在`genres`为空字符串的行。

In [ ]:
cleaned_titles = cleaned_titles.query('genres!=""')
cleaned_titles.query('genres==""')

In [ ]:
cleaned_titles = cleaned_titles.dropna(subset=["genres"])
cleaned_titles

##### 由于缺失分析所需的核心数据genres，我们将把这些观察值删除，下面检查这些空值有没有被删干净

In [ ]:
cleaned_titles.query("imdb_score.isnull()")

In [ ]:
cleaned_titles.query("genres.isnull()")

##### 删干净了，下面进行数据格式转换
##### release_year数据类型应为日期，seasons数据类型应是整数，imdb_votes数据类型应是整数¶

In [ ]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"],format='%Y')
cleaned_titles["release_year"]
#format='%Y' 告诉解析函数：输入的字符串是一个四位数的年份。

In [ ]:
cleaned_titles["seasons"] = cleaned_titles["seasons"].astype("Int64")
#包含 NaN 值，而 astype(int) 无法将 NaN 转换为整数，因此抛出错误。使用 pandas 的可空整数类型 Int64（带大写 I），它支持 NaN 值：
cleaned_titles.sample(10)

In [ ]:
cleaned_titles["imdb_votes"] = cleaned_titles["imdb_votes"].astype("Int64")
cleaned_titles.sample(10)

##### 下一步对cleaned_credits数据干净度分析

In [ ]:
cleaned_credits.info()

##### 其中character有缺失值，这个不影响，不管
##### person_id的数据类型应为字符串

In [ ]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype("str")
cleaned_credits["person_id"]

#### 处理重复数据

In [ ]:
cleaned_credits.duplicated().sum()

In [ ]:
cleaned_titles.duplicated().sum()

#### 处理不一致数据（不同值指代同一个事物）

##### 针对`cleaned_titles`，不一致数据可能存在于`genres`和`character`变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [ ]:
cleaned_titles['genres'].value_counts()

##### 从上面看出，genres列里并不存在不一致数据，各个值都在指代不同的流派。

In [ ]:
cleaned_titles['production_countries'].value_counts()

##### 由于`value_counts`执行结果中有太多值，Pandas只会默认显示开头和结尾的一些值。要完整展示结果，可以把`display.max_rows`设置为`None`，即取消展示行数上限。但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_context`，只更改临时上限。

In [ ]:
with pd.option_context("display.max_rows",None):
    print(cleaned_titles['production_countries'].value_counts())

##### 从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个的Lebanon值。
##### Lebanon的国家代码是LB，出现了39次，说明此处数据不一致。LB和Lebanon都在表示同一国家，需要进行统一。

##### 把cleaned_titles里，production_countries的"LB"和"Lebanon"统一为LB，并检查替换后是否还存在"LB"

In [ ]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].replace({"Lebanon":"LB"})
#print("Lebanon" in cleaned_titles.values)#里面有缺失值，不能用in,老实找吧
with pd.option_context("display.max_rows",None):
    print(cleaned_titles['production_countries'].value_counts())

##### 针对cleaned_credits，不一致数据可能存在于role中，我们将查看是否存在多个不同值指代同一演职员类型的情况

In [ ]:
cleaned_credits['role'].value_counts()

##### 从以上输出结果来看，role只有两种可能的值，ACTOR或DIRECTOR，不存在不一致数据。
##### 我们可以把这列的类型转换为category，好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [ ]:
cleaned_credits['role'] = cleaned_credits['role'].astype("category")
cleaned_credits['role']

#### 处理无效或错误数据

##### 可以通过DataFrame的describe方法，对数值统计信息进行快速了解。

In [ ]:
cleaned_titles.describe()

##### 没什么问题，cleaned_credits没有数值的变量，不用查

## 整理数据

In [ ]:
cleaned_titles

In [ ]:
cleaned_credits

##### 此次数据分析目标是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。
##### 那为了能同时获得流派与演员数据，我们需要把cleaned_titles和cleaned_credits，通过id作为键进行连接，因为两个数据表中id都是影视作品ID。

In [ ]:
credits_with_titles = pd.merge(cleaned_titles,cleaned_credits,on="id",how="inner")
credits_with_titles.head(10)

##### 由于我们只对挖掘演员的参演作品口碑感兴趣，导演不在我们的分析范围内，因此根据role，筛选出类型为ACTOR的观察值，供后续分析。

In [ ]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

##### 为了挖掘出各个流派中的高IMDB评分作品演员，我们需要先根据流派和演员进行分组。
##### 对演员进行分组的时候，选择的是用person_id而不是name变量，原因是名字容易出现错拼或者重名的情况，演职员ID会比演员姓名更加准确地反映是哪位

In [ ]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres","person_id"])#genres是外层索引，person_id是内层索引

##### 分组后，我们只需要对imdb_score的值进行聚合计算，因此只提取imdb_score变量，然后调用mean，来计算各个流派影视作品中，每位演员参演作品的平均IMDB评分。

In [ ]:
groupby_genres_and_person_id_imdb_score = groupby_genres_and_person_id["imdb_score"].mean()
groupby_genres_and_person_id_imdb_score

##### 我们可以调用reset_index，对层次化索引进行重置，得到更加规整的DataFrame。

In [ ]:
groupby_genres_and_person_id_imdb_score_df = groupby_genres_and_person_id_imdb_score.reset_index()
groupby_genres_and_person_id_imdb_score_df
#对 Series 操作时，reset_index() 会将其转换为 DataFrame（原索引为一列，原值为另一列）,两列索引，一列值

##### 现在针对流派和演员分组的IMDB评分数据已经整理好，可以进入后续的分析步骤了。

In [ ]:
genres_max_scores = groupby_genres_and_person_id_imdb_score_df.groupby("genres")["imdb_score"].max()
genres_max_scores

In [ ]:
genres_max_scores_with_person_id = pd.merge(genres_max_scores,groupby_genres_and_person_id_imdb_score_df,on=["genres","imdb_score"])
genres_max_scores_with_person_id

In [ ]:
person_id_with_name = cleaned_credits[["person_id","name"]].drop_duplicates()
person_id_with_name.sample(10)

In [ ]:
pd.merge(genres_max_scores_with_person_id,person_id_with_name,on="person_id")